src/core/regimes.rkt

```diff
@@ -64,26 +66,24 @@
   (set-disjoint? crit-vars non-crit-vars))
 
 ;; Requires that prog is a λ expression
-(define (all-critical-subexpressions prog)
+(define (all-critical-subexpressions expr ctx)
   (define (subexprs-in-expr expr)
     (cons expr (if (list? expr) (append-map subexprs-in-expr (cdr expr)) '())))
-  (define prog-body (program-body prog))
-  ;; We append program-variables here in case of (λ (x y) 0) or
-  ;; similar, where the variables do not appear in the body but are
-  ;; still worth splitting on
-  (for/list ([expr (remove-duplicates (append (program-variables prog)
-                                              (subexprs-in-expr prog-body)))]
+  ;; We append all variables here in case of (λ (x y) 0) or similar,
+  ;; where the variables do not appear in the body but are still worth
+  ;; splitting on
+  (for/list ([expr (remove-duplicates (append (context-vars ctx) (subexprs-in-expr expr)))]
              #:when (and (not (null? (free-variables expr)))
-                         (critical-subexpression? prog-body expr)))
+                         (critical-subexpression? expr expr)))
```

Comment:
I believe these two `expr`s should be different. Based on the diff, you should be comparing `expr` bound in the for loop on line 75 to the `expr` bound as argument for `all-critical-subexpressions`.

Reply:
Good catch